streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractTransactionalStore.java

```diff
@@ -0,0 +1,221 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.List;
+import java.util.Map;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.StateStoreContext;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+
+public abstract class AbstractTransactionalStore implements KeyValueStore<Bytes, byte[]> {
+    static final String PREFIX = "transactional-";
+    private final ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
+
+    private Path commitMarker;
+
+    final KeyValueSegment tmpStore;
+
+    Map<String, Object> configs;
+    File stateDir;
+
+    abstract KeyValueStore<Bytes, byte[]> mainStore();
+
+    AbstractTransactionalStore(final KeyValueSegment tmpStore) {
+        this.tmpStore = tmpStore;
+    }
+
+    @Deprecated
+    @Override
+    public void init(final ProcessorContext context, final StateStore root) {
+        if (context instanceof StateStoreContext) {
+            init((StateStoreContext) context, root);
+        } else {
+            throw new UnsupportedOperationException(
+                "Use TransactionalKeyValueStore#init(StateStoreContext, StateStore) instead."
+            );
+        }
+    }
+
+    @Override
+    public void init(final StateStoreContext context, final StateStore root) {
+        doInit(context.appConfigs(), context.stateDir());
+        mainStore().init(context, root);
+    }
+
+    void doInit(final Map<String, Object> configs, final File stateDir) {
+        this.configs = configs;
+        this.stateDir = stateDir;
+        commitMarker = stateDir.toPath().resolve(PREFIX + name() + ".commit");
+        tmpStore.openDB(configs, stateDir);
+    }
+
+    @Override
+    public void close() {
+        tmpStore.close();
+        mainStore().close();
+    }
+
+    @Override
+    public void commit(final Long changelogOffset) {
+        tmpStore.commit(changelogOffset);
+        createCommitMarker(changelogOffset);
+        doCommit();
+    }
+
+    @Override
+    public Long recover(final Long changelogOffset) {
+        if (Files.exists(commitMarker)) {
```

Comment:
If the marker file does exist but is corrupted and hence we cannot read out the long value, do we really want to return `null` and let the caller to wipe out the whole store?

If this happens, should we be certain that we fail at "   2. Create a commit marker with a changelog offset corresponding to the state we are committing." and hence we know nothing from the tmp store is written, so we can just safely return the changelog offset?

Reply:
This is another comment: it seems that in the current proposal, we do not necessarily guarantee that the returned offset is always strictly reflecting the store's current snapshot, and lays on the hope that since restoration is idempotent, all we need to make sure is that non committed data are not written to the store.

In that case, I'm actually wondering if the txn marker, plus the `recover()` function is necessary: since we do this tmp -> actual store copying after the whole EOS commit transaction completes, then upon failures we know that our actual store's image is definitely no more than the already-committed txn's changelog offset, so we can always just blindly start restoring from the checkpointed offset to the log-end offset (in read committed mode of course). Right?